magnolia/shared/src/main/scala-dotty/zio/config/magnolia/Descriptor.scala

```diff
@@ -0,0 +1,165 @@
+                )
+          ), desc.keys
+        ) :: summonDescriptorForCoProduct[ts]
+
+  inline def summonDescriptorAll[T <: Tuple]: List[Descriptor[_]] =
+    inline erasedValue[T] match
+      case _ : EmptyTuple => Nil
+      case _: (t *: ts) => summonInline[Descriptor[t]] :: summonDescriptorAll[ts]
+
+  inline def labelsToList[T <: Tuple]: List[String] =
+    inline erasedValue[T] match
+      case _: EmptyTuple => Nil
+      case _ : ( t *: ts) => constValue[t].toString :: labelsToList[ts]
+
+  inline given derived[T](using m: Mirror.Of[T]): Descriptor[T] =
+    inline m match
+      case s: Mirror.SumOf[T] =>
+        sum[T](using s)
+
+      case p: Mirror.ProductOf[T] =>
+        product[T](using p)
+
+  inline def sum[T](using m: Mirror.SumOf[T]) =
+    val nameOfT =
+      constValue[m.MirroredLabel]
+
+    val alternateNamesOfEnum: List[String] =
+      Macros.nameAnnotations[T].map(_.name) ++ Macros.namesAnnotations[T].flatMap(_.names)
+
+    val subClassNames =
+      labelsToList[m.MirroredElemLabels]
+
+    val subClassDescriptions =
+      summonDescriptorForCoProduct[m.MirroredElemTypes]
+
+    mergeAllProducts(
+      subClassDescriptions.map(castTo[Descriptor[T]]),
+      subClassNames
+    )
+
+  def mergeAllProducts[T](
+    allDescs: => List[Descriptor[T]],
+    subClassNames: List[String]
+  ): Descriptor[T] =
+    Descriptor(allDescs.zip(subClassNames).map({case (d, n) => {
+      if(d.isObject) d.desc else nested(n)(d.desc)
+    }}).reduce(_ orElse _))
+
+  inline def product[T](using m: Mirror.ProductOf[T]) =
+    val nameOfT =
+      constValue[m.MirroredLabel]
+
+    val fieldNames =
+      labelsToList[m.MirroredElemLabels]
+
+    mergeAllFields(
+      summonDescriptorAll[m.MirroredElemTypes],
+      fieldNames,
+      List(nameOfT),
+      lst => m.fromProduct(Tuple.fromArray(lst.toArray[Any])),
+      castTo[Product](_).productIterator.toList
+    )
+
+  def mergeAllFields[T](
+     allDescs: => List[Descriptor[_]],
+     fieldNames: => List[String],
+     namesOfClass: => List[String],
+     f: List[Any] => T,
+     g: T => List[Any],
+   ): Descriptor[T] =
+       if fieldNames.isEmpty then
+         val tryAllPaths = namesOfClass.map(n => DerivationUtils.constantString(n)).reduce(_ orElse _)
+         Descriptor(tryAllPaths.transform[T](_ => f(List.empty[Any]), t => namesOfClass.headOption.getOrElse(t.toString)))
+       else
+         val listOfDesc =
+           fieldNames.zip(allDescs).map({ case (a, b) => nested(a)(castTo[ConfigDescriptor[Any]](b.desc)) })
+
+         val descOfList =
+           collectAll(listOfDesc.head, listOfDesc.tail: _*)
+
+         Descriptor(descOfList.transform[T](f, g), fieldNames)
+
+  def castTo[T](a: Any): T =
+    a.asInstanceOf[T]
+
+  extension[E, A](e: Either[E, A]) {
+    def mapError(f: E => String) =
+      e.swap.map(f).swap
+  }
+}
```

Comment:
Casting involved can be avoided if we go via the approach similar to 
https://github.com/afsalthaj/scala3-generic/blob/master/core/src/main/scala/com/thaj/scala/three/generic/typeclasses/MapperF.scala

However, after dealing with fairly non-idiomatic scala3 way, if by any chance we end up casting again, the whole exercise will be futile, resulting in a complex implementation to exist for particularly no reason.

Example: The following cannot be implemented in scala3 without `asInstanceOf`


```scala
  def sequence(s: TupleMap[F, T]): F[TupleInversMap[F, T]] = ???
```